14 feature fe upload code

ashan36 opened the pull request:
## Done
- Basic code-upload page layout
- Implemented custom rich text editor
- Toggle buttons for toolbar
- Code block, blockquote, unordered list formatted block text
- Code block key commands
- Code block syntax highlighting based on language selector
- Submit button that pulls texteditor content out as JSON to send to the server

## To do
- Turns out the best way to bundle a configured prismjs file is to use babel-loader. In order to configure our build options, we'd need to run the eject script that comes from create-react-app. -- - - Alternatively we could just run it not as a dependency and package the entire thing up in our build. The whole thing is 44kb with the languages currently on our list.
- Set language drop down options based on user experience object.

ashan36 committed:
Created code upload page skeleton

ashan36 committed:
Added Basic Styling

ashan36 committed:
Merge branch 'dev' into 14-feature-fe-upload-code

ashan36 committed:
Added basic editor functionality

ashan36 committed:
Merge branch 'dev' into 14-feature-fe-upload-code

ashan36 committed:
Merge branch 'dev' into 14-feature-fe-upload-code

ashan36 committed:
Added code block formatting, syntax highlighting, and a key press handler

ashan36 committed:
Extracted toolbar from text editor component.

ashan36 committed:
Added submit button and submit function

Also fixed a few bugs with the functionality of the editor. The experience is a bit more streamlined now.

ashan36 committed:
Added error messaging when submitting blank editor

ashan36 committed:
Updated inline-code style to match code-block

ashan36 commented:
If you are testing the syntax highlighting, only the grammar for Javascript is loaded by default in the dependency. If you want the others you will have to download and replace prismjs with the one from the link here: https://prismjs.com/download.html#themes=prism-okaidia&languages=markup+css+clike+javascript+c+csharp+cpp+java+python+ruby

rkerins91 commented on client/src/pages/CodeUpload.js:
```diff
@@ -0,0 +1,176 @@
+const useStlyes = makeStyles({
+  root: {
+    padding: "5%"
+  },
```
I think the background here looks out of place from the rest of out site. Can add `backgroundImage: "linear-gradient(to top, rgba(87,40,184,1) 25%, rgba(119,140,255,1) 100%)",` to root to get the gradient and make the site look uniform. Also probably want to have the corners rounded like in Nick's SignUpContainer component.

ashan36 replied:
Rounded the corners.

I chose that background color because I assumed from the RFP that the main site color would be this (see profile, adding credits, and reviews mockups), and only the onboarding, login, and signup pages have the purple gradient. We can discuss more on the site styling in standup.

ashan36 committed:
Merge branch 'dev' into 14-feature-fe-upload-code

ashan36 committed:
Updated material-ui libraries due to compile errors. Fixed Grid property

ashan36 committed:
Created a simple config file to use where we can store public config variables and to reference private .env variables

ashan36 committed:
Created an stackable Snackbar component

Wrote it so it can be used in most situations across the site where we need to provide information to the user.

ashan36 committed:
Merge pull request #34 from hatchways/be-server-config-file

Created a simple config file to use where we can store public config …

ashan36 committed:
Updated upload page to add languages to the drop down based on the stored user object

Also added to the utility a map of language names mapped to their prismjs  grammar name

skassam21 commented on client/src/App.css:
```diff
@@ -31,3 +31,17 @@
+  font-size: 16px;
+  padding: 20px;
+  margin-bottom: 20px;
+} */
```
Remove this commented out code

skassam21 commented on client/src/App.js:
```diff
@@ -2,11 +2,13 @@ import React, { useState, useMemo } from "react";
 import { MuiThemeProvider } from "@material-ui/core";
 import { BrowserRouter, Route } from "react-router-dom";
 import { UserContext } from "./context/UserContext";
+import { SnackbarProvider } from "notistack";
```
Why are we using a snackbar library when material-ui has snackbars already - this will add unnecessary bloat to the application

ashan36 replied:
notistack is a wrapper for the material-ui snackbar component that allows for nicely displayed stackable snackbars. It saves a lot of customizing work to get multiple snackbars playing nice with eachother ourselves.

skassam21 commented on client/src/components/TextEditor/TextEditor.js:
```diff
@@ -0,0 +1,224 @@
+import "prismjs/themes/prism-okaidia.css";
+
+const TextEditor = props => {
+  const useStlyes = makeStyles({
```
misspelled useStyles

skassam21 commented on client/src/components/TextEditor/TextEditor.js:
```diff
@@ -0,0 +1,224 @@
+import PrismDecorator from "draft-js-prism";
+import "prismjs/themes/prism-okaidia.css";
+
+const TextEditor = props => {
```
I generally like to destructure props, but that's a preference, some people don't like to do that

skassam21 commented on client/src/components/TextEditor/TextEditor.js:
```diff
@@ -0,0 +1,224 @@
+      style === "BOLD" ||
+      style === "ITALIC" ||
+      style === "UNDERLINE" ||
+      style === "CODE"
```
These strings should be constants, possibly at the top of your file (outside the component)

skassam21 commented on client/src/components/TextEditor/TextEditor.js:
```diff
@@ -0,0 +1,224 @@
+      textShadow: "0 1px rgba(0, 0, 0, 0.3)",
+      color: "#f8f8f2"
+    }
+  };
```
This probably doens't need to be defined in the component itself

skassam21 commented on client/src/components/TextEditor/TextEditor.js:
```diff
@@ -0,0 +1,224 @@
+
+  const decorator = new PrismDecorator({
+    prism: Prism,
+    defaultSyntax: languageState
```
why is languageState a state variable... it functions the same as props.selectedLanguage?

skassam21 replied:
Also, I wonder if this needs to be keep redefined (I don't know how costly it is to keep redefining it), every time the component renders

skassam21 commented on client/src/components/TextEditor/components/SubmitButton.js:
```diff
@@ -0,0 +1,30 @@
+const SubmitButton = props => {
+  const useStyles = makeStyles({
+    button: {
+      backgroundColor: "#43DDC1",
```
Why do we have this specific custom SubmitButton - background colours for buttons generally should be coming from the theme

skassam21 commented on client/src/components/TextEditor/components/Toolbar.js:
```diff
@@ -0,0 +1,87 @@
+
+  useEffect(() => {
+    setBlockFormatButtonState(props.BlockStyle);
+  }, [props.BlockStyle]);
```
Again, I think this is overusing state - you can just use the props values

skassam21 commented on client/src/pages/CodeUpload.js:
```diff
@@ -0,0 +1,200 @@
+          <TextEditor
+            selectedLanguage={requestLanguage}
+            onSubmit={handleSubmit}
+            didSubmit={submitState}
```
The flow of submitting is confusing and hard to follow, it could probably be a bit better designed

skassam21 reviewed (commented):
I know this is merged already, but I thought I'd give my feedback for things you can think about going forward
